x/upgrade/internal/keeper/keeper.go

```diff
@@ -0,0 +1,112 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/exported"
+	"github.com/cosmos/cosmos-sdk/x/upgrade/internal/types"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+type Keeper struct {
+	storeKey        sdk.StoreKey
+	cdc             *codec.Codec
+	upgradeHandlers map[string]types.UpgradeHandler
+	haveCache       bool
+}
+
+var _ exported.Keeper = (*Keeper)(nil)
+
+// NewKeeper constructs an upgrade Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec) *Keeper {
+	return &Keeper{
+		storeKey:        storeKey,
+		cdc:             cdc,
+		upgradeHandlers: map[string]types.UpgradeHandler{},
+	}
+}
+
+// SetUpgradeHandler sets an UpgradeHandler for the upgrade specified by name. This handler will be called when the upgrade
+// with this name is applied. In order for an upgrade with the given name to proceed, a handler for this upgrade
+// must be set even if it is a no-op function.
+func (k *Keeper) SetUpgradeHandler(name string, upgradeHandler types.UpgradeHandler) {
+	k.upgradeHandlers[name] = upgradeHandler
+}
+
+// ScheduleUpgrade schedules an upgrade based on the specified plan
+func (k *Keeper) ScheduleUpgrade(ctx sdk.Context, plan types.Plan) sdk.Error {
```

Comment:
From my understanding, does this function allow an already existing plan to have its duedate updated if the plan has not already been completed?

Reply:
You are right. As stated, it will allow you to overwrite the existing plan.
I think that is not so good actually, and will update it to fail if one exists.

You would have to explicitly cancel (`ClearUpgradePlan`) and then Schedule a new one